File: frontend/src/services/userService.ts


export interface Role {
  id: number;
  name: string;
}

export interface UserRole {
  roleId: number;
  role: Role;
}

import api from './api';

export interface User {
  id: string;
  username: string;
  email: string;
  departmentId?: number;
  department?: string;
  roles: string[];
  roleIds?: number[];
  status: 'Active' | 'Locked';
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
}

export const userService = {
  getUsers: async (page = 1, pageSize = 10, departmentId?: number, search?: string): Promise<PagedResult<User>> => {
    const params = new URLSearchParams({
      pageNumber: page.toString(),
      pageSize: pageSize.toString()
    });

    if (departmentId) params.append('departmentId', departmentId.toString());
    if (search) params.append('searchTerm', search);

    const data = (await api.get<any>(`/users?${params.toString()}`)).data;

    return {
      ...data,
      items: data.items.map((u: any) => ({
        id: u.id,
        username: u.username,
        email: u.email,
        departmentId: u.departmentId,
        department: u.department?.name,
        roles: u.roles?.map((r: any) => r.name) || [],
        roleIds: u.roles?.map((r: any) => r.id) || [],
        status: u.isActive ? 'Active' : 'Locked'
      }))
    };
  },
  updateUser: async (user: User): Promise<User> => {

    const payload = {
      id: user.id,
      username: user.username,
      email: user.email,
      departmentId: user.departmentId,
      isActive: user.status === 'Active',
      roleIds: user.roleIds
    };
    await api.patch(`/users/${user.id}`, payload);
    return user;
  },
  createUser: async (payload: any): Promise<User> => {
    const data = (await api.post<any>('/users', payload)).data;
    return {
      id: data.id,
      username: data.username,
      email: data.email,
      departmentId: data.departmentId,
      department: data.department?.name,
      roles: data.roles?.map((r: any) => r.name) || [],
      roleIds: data.roles?.map((r: any) => r.id) || [],
      status: data.isActive ? 'Active' : 'Locked'
    };
  }
};

File: backend/Application/Services/UserService.cs
using Application.DTOs;
using Application.Interfaces;
using Domain.Interfaces;
using Domain.Entities;

namespace Application.Services;

public class UserService : IUserService
{
    private readonly IUserRepository _repository;
    private readonly IRoleRepository _roleRepository;
    private readonly IDepartmentRepository _departmentRepository;

    public UserService(
        IUserRepository repository,
        IRoleRepository roleRepository,
        IDepartmentRepository departmentRepository)
    {
        _repository = repository;
        _roleRepository = roleRepository;
        _departmentRepository = departmentRepository;
    }

    public async Task<PagedResult<UserDto>> GetPagedUsersAsync(UserQueryParameters query)
    {
        var (items, totalCount) = await _repository.GetPagedAsync(
            query.PageNumber,
            query.PageSize,
            query.DepartmentId,
            query.SearchTerm);

        return new PagedResult<UserDto>
        {
            Items = items.Select(MapToDto).ToList(),
            TotalCount = totalCount,
            PageNumber = query.PageNumber,
            PageSize = query.PageSize
        };
    }

    public async Task<UserDto?> GetUserByIdAsync(Guid id)
    {
        var user = await _repository.GetByIdAsync(id);
        if (user == null) return null;
        return MapToDto(user);
    }

    public async Task<UserDto> CreateUserAsync(CreateUserDto dto)
    {
        var user = new User
        {
            Id = Guid.NewGuid(),
            Username = dto.Username,
            Email = dto.Email,
            PasswordHash = dto.Password,
            DepartmentId = dto.DepartmentId,
            IsActive = true
        };

        if (dto.RoleIds != null && dto.RoleIds.Any())
        {
            foreach (var roleId in dto.RoleIds)
            {
                user.UserRoles.Add(new UserRole { UserId = user.Id, RoleId = roleId });
            }
        }

        await _repository.AddAsync(user);

        var savedUser = await _repository.GetByIdAsync(user.Id);
        return MapToDto(savedUser!);
    }

    public async Task<UserDto?> UpdateUserAsync(Guid id, UpdateUserDto dto)
    {
        var user = await _repository.GetByIdAsync(id);
        if (user == null) return null;

        if (dto.Email != null) user.Email = dto.Email;
        if (dto.DepartmentId.HasValue) user.DepartmentId = dto.DepartmentId.Value;
        if (dto.IsActive.HasValue) user.IsActive = dto.IsActive.Value;

        user.UpdatedAt = DateTime.UtcNow;
        await _repository.UpdateAsync(user);

        if (dto.RoleIds != null)
        {
            await _repository.UpdateUserRolesAsync(user.Id, dto.RoleIds);
            
            user = await _repository.GetByIdAsync(user.Id);
        }

        return MapToDto(user!);
    }

    public async Task<bool> DeleteUserAsync(Guid id)
    {
        var user = await _repository.GetByIdAsync(id);
        if (user == null) return false;

        await _repository.DeleteAsync(id);
        return true;
    }

    public async Task<bool> AssignRoleAsync(Guid userId, int roleId)
    {
        var user = await _repository.GetByIdAsync(userId);
        var role = await _roleRepository.GetByIdAsync(roleId);

        if (user == null || role == null) return false;

        if (!user.UserRoles.Any(ur => ur.RoleId == roleId))
        {
            user.UserRoles.Add(new UserRole { UserId = userId, RoleId = roleId });
            await _repository.UpdateAsync(user);
        }
        return true;
    }

    public async Task<bool> RemoveRoleAsync(Guid userId, int roleId)
    {
        var user = await _repository.GetByIdAsync(userId);
        if (user == null) return false;

        var userRole = user.UserRoles.FirstOrDefault(ur => ur.RoleId == roleId);
        if (userRole != null)
        {
            user.UserRoles.Remove(userRole);
            await _repository.UpdateAsync(user);
        }
        return true;
    }

    private UserDto MapToDto(User user)
    {
        return new UserDto
        {
            Id = user.Id,
            Username = user.Username,
            Email = user.Email,
            DepartmentId = user.DepartmentId,
            IsActive = user.IsActive,
            CreatedAt = user.CreatedAt,
            UpdatedAt = user.UpdatedAt,
            Department = user.Department != null ? new DepartmentDto { Id = user.Department.Id, Name = user.Department.Name } : null,
            Roles = user.UserRoles?.Select(ur => new RoleDto { Id = ur.Role.Id, Name = ur.Role.Name }).ToList() ?? new List<RoleDto>()
        };
    }
}
